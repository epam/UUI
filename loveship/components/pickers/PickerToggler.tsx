import * as React from 'react';
import * as types from '../types';
import colorStyle from '../../assets/styles/scss/loveship-color-vars.scss';
import { PickerToggler as UuiPickerToggler, PickerTogglerProps } from '@epam/uui-components';
import { DataRowProps } from '@epam/uui-core';
import { TextPlaceholder } from '../typography';
import { systemIcons } from '../icons/icons';
import { Tag, TagSize } from '../widgets';
import css from './PickerToggler.scss';

const defaultSize = '36';
const defaultMode = 'form';

const mapSize = {
    '48': '42',
    '42': '36',
    '36': '30',
    '30': '24',
    '24': '18',
    none: 'none',
} as const;

export interface PickerTogglerMods extends types.EditMode {
    size?: types.ControlSize;
}

function applyPickerTogglerMods(mods: PickerTogglerMods) {
    return [colorStyle.colorSky, css.root, css['size-' + (mods.size || defaultSize)], css['mode-' + (mods.mode || defaultMode)]];
}

function PickerTogglerComponent<TItem extends string, TId>(
    props: PickerTogglerProps<TItem, TId> & PickerTogglerMods,
    ref: React.ForwardedRef<HTMLElement>
) {
    const getCaption = (row: DataRowProps<TItem, TId>) => {
        const maxItems = props.maxItems || props.maxItems === 0 ? props.maxItems : 100;

        if (row.isLoading) {
            return <TextPlaceholder />;
        } else if (!props.getName || props.selection?.length > maxItems) {
            return row.value;
        } else {
            return props.getName(row.value);
        }
    };

    const renderItem = (row: DataRowProps<TItem, TId>) => {
        const tagSize = mapSize[props.size] as TagSize;

        return (
            <Tag
                key={row.rowKey}
                caption={getCaption(row)}
                color="night300"
                tabIndex={-1}
                size={props.size ? tagSize : '30'}
                onClear={e => {
                    row.onCheck?.(row);
                    e.stopPropagation();
                }}
                isDisabled={props.isDisabled || props.isReadonly || row?.checkbox?.isDisabled}
            />
        );
    };

    return (
        <UuiPickerToggler
            {...props}
            ref={ref}
            isDropdown={props.isDropdown && !props.minCharsToSearch}
            cx={[applyPickerTogglerMods(props), props.cx]}
            renderItem={!!props.renderItem ? props.renderItem : renderItem}
            getName={item => (props.getName ? props.getName(item) : item)}
            cancelIcon={systemIcons[props.size || defaultSize].clear}
            dropdownIcon={systemIcons[props.size || defaultSize].foldingArrow}
        />
    );
}

export const PickerToggler = React.forwardRef(PickerTogglerComponent) as <TItem, TId>(
    props: PickerTogglerProps<TItem, TId>,
    ref: React.ForwardedRef<HTMLElement>
) => JSX.Element;
