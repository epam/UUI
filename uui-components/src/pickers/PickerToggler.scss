.container {
    cursor: pointer;

    &:global(.uui-input-box) {
        display: flex;
        align-items: center;
        justify-content: space-between;
        outline: none;
        position: relative;
    }

    :global(.uui-input) {
        flex: 1 0 auto;
        min-width: 0;
        border: 0;
        outline: none;
        background: none;
        cursor: pointer;

        &.single-input {
            flex: 1 1 auto;
        }
    }
}

.body {
    display: flex;
    width: 100%;
    overflow: hidden;
    outline: none;
    align-items: center;

    &.multiline {
        flex-wrap: wrap;
    }
}

.actions {
    display: flex;
    flex-shrink: 0;
    position: absolute;
    right: 6px;
}

.toggler {
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
    cursor: pointer;
}

.hidden {
    visibility: hidden;
}