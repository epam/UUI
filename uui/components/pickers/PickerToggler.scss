@use  '../../assets/styles/index' as *;

$shadow: 0 1px 3px 0 rgba(29, 30, 38, 0.15);

@mixin picker-toggler-vertical-layout($size, $borderWidth) {
    &:global(.uui-input-box) {
        min-height: $size;
        border-width: $borderWidth;
    }

    :global(.uui-input) {
        @include text-size($size);

        @if ($size == 24px) {
            padding-top: 0;
            padding-bottom: 0;
        } @else if ($size == 30px) {
            padding-top: 3px;
            padding-bottom: 3px;
        } @else if ($size == 36px or $size == 42px) {
            padding-top: 6px;
            padding-bottom: 6px;
        } @else if ($size == 48px) {
            padding-top: 9px;
            padding-bottom: 9px;
        }
    }

    :not(:global(.uui-button-box)) > :global(.uui-icon) {
        height: $size - 6px;
        margin-bottom: 2px;
    }
}

.root {
    padding-top: 2px;
    padding-left: 5px;
    padding-right: 5px;
    width: 100%;
    box-sizing: border-box;
    border-style: solid;

    &.mode-form {
        background-color: var(--uui-picker-toggler-bg);
        border-color: var(--uui-picker-toggler-border);
        fill: var(--uui-picker-toggler-icon);
        border-radius: var(--uui-picker-toggler-border-radius);

        :global(.uui-icon):global(.-clickable), :global(.uui-icon-dropdown) {
            &:global(.uui-enabled) {
                fill: var(--uui-picker-toggler-icon-clickable);

                &:hover {
                    fill: var(--uui-picker-toggler-icon-clickable-hover);
                }
            }

            &:global(.uui-disabled) {
                fill: var(--uui-picker-toggler-icon-disabled);
            }
        }

        &:hover {
            border-color: var(--uui-picker-toggler-border-hover);
        }

        &:global(.uui-focus) {
            border-color: var(--uui-picker-toggler-border-active);
            outline: none;
        }
    }

    &.mode-cell {
        background-color: transparent;
        border-color: transparent;
        fill: var(--uui-picker-toggler-icon);

        :global(.uui-icon):global(.-clickable), :global(.uui-icon-dropdown) {
            cursor: pointer;

            &:global(.uui-enabled) {
                fill: transparent;
            }

            &:global(.uui-disabled) {
                fill: transparent;
            }
        }

        &:hover {
            background-color: white;
            border-color: var(--uui-picker-toggler-border);

            :global(.uui-icon):global(.-clickable), :global(.uui-icon-dropdown) {
                &:global(.uui-enabled) {
                    fill: var(--uui-picker-toggler-icon-clickable);

                    &:hover {
                        fill: var(--uui-picker-toggler-icon-clickable-hover);
                    }
                }

                &:global(.uui-disabled) {
                    fill: transparent;
                }
            }
        }

        &:global(.uui-focus) {
            background-color: white;
            border-color: var(--uui-picker-toggler-border-active);
            outline: none;
        }
    }

    &:global(.uui-invalid) {
        &.mode-form, &.mode-cell {
            border-color: var(--uui-picker-toggler-border-error);
        }
    }

    &:global(.uui-readonly) {
        @include readonly-input();

        :global(.uui-icon-dropdown), :global(.uui-icon-cancel) {
            visibility: hidden;
        }
    }

    &:global(.uui-disabled) {
        @include disabled-input();
        fill: var(--uui-picker-toggler-icon-disabled);

        :global(.uui-input) {
            cursor: default;

            &::placeholder {
                color: var(--uui-picker-toggler-text-disabled);
            }
        }

        :global(.uui-icon-dropdown), :global(.uui-icon-cancel) {
            visibility: hidden;
        }
    }

    :global(.uui-input) {
        padding-left: 6px;
        padding-right: 6px;
        margin-bottom: 2px;
        font-family: var(--font-regular);
        color: var(--uui-picker-toggler-text);

        &::placeholder {
            color: var(--uui-picker-toggler-placeholder);
        }

        &::-moz-placeholder {
            opacity: 1;
        }
    }

    :global(.uui-caption) {
        @include overflow-ellipsis();
        font-family: var(--font-regular);
        align-self: center;
        white-space: nowrap;
    }

    :global(.uui-button-box) {
        margin-right: 3px;
        margin-bottom: 2px;

        &:last-of-type {
            margin-right: 0;
        }
    }

    :global(.uui-placeholder) {
        color: var(--uui-picker-toggler-text);
    }

    :global(.uui-placeholder)::placeholder {
        color: var(--uui-picker-toggler-text);
    }

    :global(.uui-placeholder)::-moz-placeholder {
        opacity: 1;
    }

    :global(.uui-icon-cancel) {
        margin-left: 6px;
    }

    // Sizes
    &.size-24 {
        @include picker-toggler-vertical-layout(24px, 1px);
        min-width: 24px;
    }

    &.size-30 {
        @include picker-toggler-vertical-layout(30px, 1px);
        min-width: 30px;
    }

    &.size-36 {
        @include picker-toggler-vertical-layout(36px, 1px);
        min-width: 36px;
    }

    &.size-42 {
        @include picker-toggler-vertical-layout(42px, 1px);
        min-width: 42px;
    }

    &.size-48 {
        @include picker-toggler-vertical-layout(48px, 1px);
        min-width: 48px;
    }
}
