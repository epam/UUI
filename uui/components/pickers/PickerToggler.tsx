import * as React from 'react';
import * as types from '../types';
import { PickerToggler as UuiPickerToggler, PickerTogglerProps } from '@epam/uui-components';
import { PickerTogglerTag, PickerTogglerTagProps } from './PickerTogglerTag';
import css from './PickerToggler.module.scss';
import { systemIcons } from '../../icons/icons';

const defaultSize = '36';
const defaultMode = types.EditMode.FORM;

export interface PickerTogglerMods extends types.IHasEditMode {
    /**
     * Defines component size
     * @default 36
     */
    size?: '24' | '30' | '36' | '42' | '48';
}

function applyPickerTogglerMods(mods: PickerTogglerMods) {
    return [
        css.root,
        css['size-' + (mods.size || defaultSize)],
        css['mode-' + (mods.mode || defaultMode)],
    ];
}

function PickerTogglerComponent<TItem extends string, TId>(props: PickerTogglerProps<TItem, TId> & PickerTogglerMods, ref: React.ForwardedRef<HTMLElement>): JSX.Element {
    const renderItem = (itemProps: PickerTogglerTagProps<TItem, TId>) => {
        const itemPropsWithSize = { ...itemProps, size: props.size };
        if (!!props.renderItem) {
            return props.renderItem(itemPropsWithSize);
        }
        return <PickerTogglerTag { ...itemPropsWithSize } />;
    };

    return (
        <UuiPickerToggler
            { ...props }
            ref={ ref }
            cx={ [applyPickerTogglerMods(props), props.cx] }
            renderItem={ renderItem }
            getName={ (item) => (props.getName ? props.getName(item) : item) }
            cancelIcon={ systemIcons.clear }
            dropdownIcon={ systemIcons.foldingArrow }
        />
    );
}

export const PickerToggler = /* @__PURE__ */React.forwardRef(PickerTogglerComponent) as <TItem, TId>(
    props: PickerTogglerProps<TItem, TId> & PickerTogglerMods,
    ref: React.ForwardedRef<HTMLElement>
) => JSX.Element;
