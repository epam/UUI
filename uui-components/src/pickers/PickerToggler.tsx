import * as React from 'react';
import { IPickerToggler, IHasIcon, IHasCX, ICanBeReadonly, Icon, uuiMod, uuiElement, uuiMarkers, DataRowProps, closest, cx, IHasRawProps } from '@epam/uui';
import { IconContainer } from '../layout';
import * as css from './PickerToggler.scss';
import { i18n } from "../../i18n";

export interface PickerTogglerProps<TItem, TId = any> extends IPickerToggler<TItem, TId>, IHasIcon, IHasCX, ICanBeReadonly, IHasRawProps<HTMLDivElement>, React.PropsWithRef<any> {
    cancelIcon?: Icon;
    dropdownIcon?: Icon;
    autoFocus?: boolean;
    renderItem?(props: DataRowProps<TItem, TId>): React.ReactNode;
    getName?: (item: DataRowProps<TItem, TId>) => string;
    entityName?: string;
    maxItems?: number;
    isSingleLine?: boolean;
    pickerMode: 'single' | 'multi';
    onKeyDown?(e: React.KeyboardEvent<HTMLElement>): void;
    onBlur?(e: React.SyntheticEvent<HTMLElement>): void;
    onFocus?(e?: React.SyntheticEvent<HTMLElement>): void;
    disableSearch?: boolean;
    disableClear?: boolean;
    minCharsToSearch?: number;
}

interface PickerTogglerState {
    inFocus?: boolean;
    isActive?: boolean;
}

export class PickerToggler<TItem, TId> extends React.Component<PickerTogglerProps<TItem, TId>, PickerTogglerState> {
    state = {
        inFocus: false,
        isActive: false,
    };

    toggleContainer: HTMLDivElement | null = null;

    componentDidMount() {
        window.document.addEventListener('click', this.handleActive);
        if (this.props.autoFocus && !this.props.disableSearch) {
            this.handleFocus();
        }
    }

    componentWillUnmount() {
        window.document.removeEventListener('click', this.handleActive);
    }

    handleChange = (e: React.ChangeEvent<HTMLInputElement>) => this.props.onValueChange && this.props.onValueChange(e.target.value);


    handleFocus = (e?: React.SyntheticEvent<HTMLElement>) => {
        this.props.onFocus && this.props.onFocus(e);
        const input = this.toggleContainer?.querySelector('input');
        input && input.focus();
        this.setState({ inFocus: true });
    }

    handleBlur = (e: React.SyntheticEvent<HTMLElement>) => {
        this.props.onBlur && this.props.onBlur(e);
        this.setState({ inFocus: false });
        const input = this.toggleContainer?.querySelector('input');
        input && input.blur();
    }

    handleActive = (e: Event) => {
        if (closest((e.target as HTMLElement), this.toggleContainer)) {
            this.setState({ isActive: true });
        }
        if (this.state.isActive && !closest((e.target as HTMLElement), this.toggleContainer)) {
            this.setState({ isActive: false });
        }
    }

    handleKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
        this.props.onKeyDown && this.props.onKeyDown(e);
    }

    handleCrossIconClick = (e: React.SyntheticEvent<HTMLElement>) => {
        this.props.onClear && this.props.onClear();
        e.stopPropagation();
    }

    renderItems() {
        let maxItems = (this.props.maxItems || this.props.maxItems === 0) ? this.props.maxItems : 100;
        if (this.props.selection && this.props.selection.length > maxItems) {
            let item = {
                value: i18n.pickerToggler.createItemValue(this.props.selection.length, this.props.entityName || ''),
                onCheck: () => this.props.onClear && this.props.onClear(),
            };
            return this.props.renderItem && this.props.renderItem(item as any);
        } else {
            return this.props.selection && this.props.selection.map(row => this.props.renderItem && this.props.renderItem(row));
        }
    }

    renderInput() {
        let isActivePlaceholder = this.props.pickerMode === 'single' && this.props.selection && !!this.props.selection[0];
        let placeholder = isActivePlaceholder ? this.props.getName(this.props.selection[0]) : this.props.placeholder;

        if (this.props.disableSearch) {
            if (this.props.pickerMode === 'multi' && this.props.selection.length > 0) {
                return null;
            }

            return <div
                aria-haspopup='listbox'
                aria-multiselectable={ this.props.pickerMode === 'multi' }
                className={ cx(
                    uuiElement.input,
                    this.props.selection.length === 0 && uuiElement.placeholder,
                    this.props.pickerMode === 'single' && css.singleInput,
                    css.toggler,
                ) }>
                { placeholder }
            </div>;
        }

        return <input
            type={ "text" }
            aria-haspopup='listbox'
            aria-multiselectable={ this.props.pickerMode === 'multi' }
            className={ cx(uuiElement.input,
                this.props.pickerMode === 'single' && css.singleInput,
                isActivePlaceholder && (!this.state.inFocus || this.props.isReadonly) && uuiElement.placeholder)
            }
            disabled={ this.props.isDisabled }
            placeholder={ placeholder }
            value={ this.props.value || '' }
            readOnly={ this.props.isReadonly }
            onChange={ this.handleChange }
        />;
    }

    togglerPickerOpened = () => {
        if (this.state.inFocus && this.props.value && !this.props.disableSearch) return;
        this.props.onClick();
    }

    render() {
        const icon = this.props.icon && <IconContainer icon={ this.props.icon } onClick={ this.props.onIconClick } />;

        return (
            <div
                onClick={ () => { this.props.onClick && !this.props.isReadonly && this.togglerPickerOpened(); this.handleFocus(); } }
                ref={ el => { this.toggleContainer = el; } }
                className={ cx(css.container,
                    uuiElement.inputBox,
                    this.props.isDisabled && uuiMod.disabled,
                    this.props.isReadonly && uuiMod.readonly,
                    this.props.isInvalid && uuiMod.invalid,
                    (!this.props.isReadonly && !this.props.isDisabled && this.props.onClick) && uuiMarkers.clickable,
                    (!this.props.isReadonly && this.state.inFocus) && uuiMod.focus,
                    (!this.props.isReadonly && this.state.isActive) && uuiMod.active,
                    this.props.cx,
                ) }
                tabIndex={ 0 }
                onFocus={ this.handleFocus }
                onBlur={ this.handleBlur }
                onKeyDown={ this.handleKeyDown }
                {...this.props.rawProps}
            >
                <div
                    className={ cx(css.body, !this.props.isSingleLine && this.props.pickerMode !== 'single' && css.multiline) }
                >
                    { this.props.iconPosition !== 'right' && icon }
                    { this.props.pickerMode !== 'single' && this.renderItems() }
                    { this.renderInput() }
                    { this.props.iconPosition === 'right' && icon }
                </div>
                <div  className={ cx(css.actions) }>
                    { !this.props.disableClear && (this.props.value || this.props.selection && this.props.selection.length > 0) && <IconContainer
                        cx={ cx('uui-icon-cancel', uuiMarkers.clickable, (this.props.isReadonly || this.props.isDisabled) && css.hidden) }
                        isDisabled={ this.props.isDisabled }
                        icon={ this.props.cancelIcon }
                        onClick={ this.handleCrossIconClick }
                    /> }
                    { this.props.isDropdown && <IconContainer
                        onClick={ (e) => { this.props.onClick(); e.stopPropagation(); }  }
                        icon={ this.props.dropdownIcon }
                        flipY={ this.props.isOpen }
                        cx={ (this.props.isReadonly || this.props.isDisabled) && css.hidden }
                    /> }
                </div>
            </div>
        );
    }
}
