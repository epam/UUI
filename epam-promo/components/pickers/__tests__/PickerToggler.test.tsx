import React from 'react';
import renderer from 'react-test-renderer';
import { PickerToggler } from '../PickerToggler';

describe('PickerToggler', () => {
    it('should be rendered correctly', () => {
        const tree = renderer.create(<PickerToggler pickerMode="single" searchPosition="none" />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly', () => {
        const tree = renderer
            .create(
                <PickerToggler
                    pickerMode="multi"
                    onClick={jest.fn}
                    onKeyDown={jest.fn}
                    value={null}
                    onValueChange={jest.fn}
                    size="48"
                    maxItems={6}
                    getName={item => item.value}
                    selection={[
                        {
                            id: 'test',
                            index: 1,
                            rowKey: 'test',
                            value: 'test',
                        },
                    ]}
                    onBlur={jest.fn}
                    onClear={jest.fn}
                    searchPosition="none"
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
