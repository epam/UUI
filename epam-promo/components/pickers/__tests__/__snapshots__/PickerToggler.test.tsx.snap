// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickerToggler should be rendered correctly 1`] = `
<div
  className="container uui-input-box root size-36 mode-form"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="body"
  >
    <input
      aria-haspopup="listbox"
      aria-multiselectable={false}
      className="uui-input singleInput"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="actions"
  />
</div>
`;

exports[`PickerToggler should be rendered correctly 2`] = `
<div
  className="container uui-input-box -clickable root size-48 mode-form"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="body multiline"
  >
    <div
      className="container uui-button-box uui-enabled root tag-color size-42 root"
      href={null}
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="container uui-icon uui-enabled -clickable -clickable"
        onClick={[Function]}
        style={Object {}}
      >
        <svg
          className=""
          height={24}
          viewBox="0 0 12 24"
          width={12}
        >
          <use
            xlinkHref="#btn-cross-24.svg"
          />
        </svg>
      </div>
    </div>
    <input
      aria-haspopup="listbox"
      aria-multiselectable={true}
      className="uui-input"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="actions"
  >
    <div
      className="container uui-icon uui-enabled -clickable uui-icon-cancel -clickable"
      onClick={[Function]}
      style={Object {}}
    >
      <svg
        className=""
        height={24}
        viewBox="0 0 12 24"
        width={12}
      >
        <use
          xlinkHref="#btn-cross-24.svg"
        />
      </svg>
    </div>
  </div>
</div>
`;
